### PR TITLE
feat: Log cache invalidations

### LINKED ISSUE #13134: Add logging for cache invalidations
### Technical TODO

After a talk with @beberlei we found that it could be useful to know, from where a cache invalidation is triggered. Often this indicates poorly written extensions or similar reasons, which are individual for each project.

An example implementation could be found in his plugin https://github.com/tideways/shopware-invalidation-logger-plugin

### PR DESCRIPTION
**Pull request by umutdogan4291**
### 1. Why is this change necessary?
This change is necessary to make cache invalidation more transparent.

### 2. What does this change do, exactly?
This change logs the cache tags that are invalidated whenever a cache invalidation happens.

### 3. Describe each step to reproduce the issue or behaviour.

1. Run a Shopware instance without this change.
2. Perform an action that triggers cache invalidation
3. Notice that no information about invalidated cache tags is logged.
4. With this change applied, the invalidated cache tags are written to the log when cache is cleared.

### 4. Please link to the relevant issues (if any).
closes #13134

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have updated developer-facing release notes if this change is **relevant** for external developers:
  - Add a short entry to `RELEASE_INFO-6.<major>.md` under “Upcoming” for informational changes, including the consequences of the change and how it affects external developers.
  - Add an `UPGRADE` section in `UPGRADE-6.<next-major>.md` for breaking changes (what/why/impact/how to adapt).
  - See the [Documenting a Release Process](https://github.com/shopware/shopware/blob/trunk/delivery-process/documenting-a-release.md) for details.
- [ ] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfilled them

### CODE REVIEW

File: RELEASE_INFO-6.7.md [hunk 1]
```diff
@@ -77,6 +77,9 @@ This affects the following exception factory methods:
 * `DataAbstractionLayerException::unexpectedFieldType(...)`
 * `WebhookException::invalidDataMapping(...)`
 * `WebhookException::unknownEventDataType(...)`
+### Logging for invalidated cache tags
```
Comment (mitelg) [inline RELEASE_INFO-6.7.md:145]:
```suggestion

### Logging for invalidated cache tags
```

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php [hunk 2]
```diff
@@ -106,11 +104,40 @@ private function purge(array $keys): void
...
     {
         return $this->requestStack->getMainRequest()?->headers->get(PlatformRequest::HEADER_FORCE_CACHE_INVALIDATE) === '1';
     }
+
+    /**
+     * @return array<string, int|string>|null
+     */
+    private function findCaller(): ?array
+    {
+        foreach (debug_backtrace(\DEBUG_BACKTRACE_IGNORE_ARGS) as $frame) {
```
Comment (mitelg) [inline src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: should we limit the backtrace somehow? so it is not trying to go back for 20 or more steps? If I remember correctly the plugin does the same

Comment (umutdogan4291) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: makes sense. we will need a few most likely. I can use the same value (10) like the plugin

Comment (mitelg) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: I am not sure, if this has some memory consumption implications. we can probably even use less than 10

Comment (umutdogan4291) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: I checked the test cases and we only use 4th level. what about setting it 5 :thinking: 

Or we can get the trace in invalidate() and pass it. We would have less trace. I didn't prefer this to not add more complexity :thinking:

Comment (mitelg) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: if 5 is also enough, then we should take that :+1:

**Review by mitelg — APPROVED** [review]

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php [hunk 3]
```diff
@@ -106,11 +104,40 @@ private function purge(array $keys): void
...
+    {
+        foreach (debug_backtrace(\DEBUG_BACKTRACE_IGNORE_ARGS, 5) as $frame) {
+            if (($frame['class'] ?? null) === self::class) {
+                continue; // skip CacheInvalidator methods
+            }
+
+            if (isset($frame['file'], $frame['class'])) {
+                return [
+                    'class' => $frame['class'],
+                    'function' => $frame['function'],
```
Comment (keulinho) [inline src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: i assume in real life most of the time this will be some method from `\Shopware\Core\Framework\Adapter\Cache\CacheInvalidationSubscriber` , but i guess that is exactly what's expected? 🤔 as it is already helpful which part of the automatic cache invalidation was triggered

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php [hunk 4]
```diff
@@ -106,11 +104,40 @@ private function purge(array $keys): void
...
+     * @return array<string, int|string>|null
+     */
+    private function findCaller(): ?array
+    {
+        foreach (debug_backtrace(\DEBUG_BACKTRACE_IGNORE_ARGS, 5) as $frame) {
+            if (($frame['class'] ?? null) === self::class) {
+                continue; // skip CacheInvalidator methods
+            }
+
+            if (isset($frame['file'], $frame['class'])) {
```
Comment (nfortier-shopware) [inline src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: what about `$frame['function']`?

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php [hunk 5]
```diff
@@ -106,11 +104,40 @@ private function purge(array $keys): void
             $this->httpCacheStore->setMultiple($list);
         }
 
+        $this->logger->info(
```
Comment (keulinho) [inline src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: wondering if we should make the log level configurable, because on info level there might be alot going on, and it might not be easy to find what you are looking for 🤔

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php [hunk 6]
```diff
@@ -106,11 +104,40 @@ private function purge(array $keys): void
...
+            if (isset($frame['file'], $frame['class'])) {
+                return [
+                    'class' => $frame['class'],
+                    'function' => $frame['function'],
+                ];
+            }
+        }
+
+        /** Defensive fallback, not reachable in normal execution. */
+        return null; // @codeCoverageIgnore
```
Comment (nfortier-shopware) [inline src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: Let's not use this annotation.

It's ok for now to not have everything covered and it can highlight something to refactor.

But also, since debug_backtrace with such depth will not allow this to happen, what's the benefit of such defensive fallback?

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php [hunk 7]
```diff
@@ -106,11 +104,40 @@ private function purge(array $keys): void
...
 
     private function shouldForceInvalidate(): bool
     {
         return $this->requestStack->getMainRequest()?->headers->get(PlatformRequest::HEADER_FORCE_CACHE_INVALIDATE) === '1';
     }
+
+    /**
+     * @return array<string, int|string>|null
+     */
+    private function findCaller(): ?array
```
Comment (AydinHassan) [inline src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: Do we maybe wanna do this lazily? Eg if it will actually be logged based on the current log level? Not sure how easy that is but it could be quite heavy doing all those backtrace calls

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php (hunk 5 above)
Comment (umutdogan4291) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: I agree that info could get noisy. If we introduce a config for the log level, Imo we can also let the users turn this logging on/off.

Is it common in shopware backend to use system configs for this kind of granular logging decisions.

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php (hunk 6 above)
Comment (nfortier-shopware) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: Would the fallback be critical, you can put return debug_backtrace(\DEBUG_BACKTRACE_IGNORE_ARGS, 5); in a protected getLastBacktraces()
to emulate it then in a test by extending CacheInvalidator (this kind of isolation has been done elsewhere)

This way findCaller would stay private, and you could reach null with simulating the backtrace given to only reach the skipped line

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php (hunk 4 above)
Comment (umutdogan4291) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: thanks :) 
I must use 'function' instead of file as we don't log 'file'

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php (hunk 6 above)
Comment (umutdogan4291) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: Thanks. I can try that because I would like to keep the fallback. It makes the return explicit and gives us a bit of safety in case this gets refactored later.

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php (hunk 3 above)
Comment (umutdogan4291) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: Yes, for now it’ll be always CacheInvalidationSubscriber, but that already gives a context

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php (hunk 7 above)
Comment (umutdogan4291) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: Afaik the logger doesn't expose the current level. If you have a trick or approach in mind I am happy to try it.

[This comment](https://github.com/shopware/shopware/pull/14253/files#r2686500058) may be helpful.

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php (hunk 5 above)
Comment (mitelg) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: I am not aware of a place, where we make the log level for certain things configurable :thinking: so, imo we should not start here :sweat_smile: 

aren't there other ways to filter the log with plugins, etc?

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php (hunk 6 above)
Comment (umutdogan4291) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: ah I see now that the class is final. I can't extend in a test case. Having it in a helper, injecting and mocking that would be an option but I feel like it's unnecessary complexity for just testing an edge case.

Also removing the fallback causes phpstan error.

Comment (mitelg) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: I would remove the annotation and we live with the uncovered line

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php (hunk 4 above)
Comment (umutdogan4291) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: I just realized that function is always set as phpstan says. So we don't need to check it.

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php (hunk 6 above)
Comment (umutdogan4291) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: I can live with it :) I added it because I assumed that it will block the PR :thinking:

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php (hunk 7 above)
Comment (AydinHassan) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: Could maybe duck type on `\Monolog\Logger` and use the `isHandling` method. I'm not entirely sure it's worth it, but I suppose `debug_backtrace` might be CPU heavy and seems a bit wasteful in prod envs.

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php (hunk 4 above)
Comment (nfortier-shopware) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: yes alright)

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php (hunk 7 above)
Comment (AydinHassan) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: Another option, similar to what Jonas suggested, just make the logging configurable eg enabled/disabled. I can imagine this is something you need when debugging an issue you can enable it, collect the logs, then disable it when you don't need it?

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php (hunk 6 above)
Comment (nfortier-shopware) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: The report is not yet hard blocking.

And yes you are right, I didn't see the annotation final, phpstan would not let that trick fly far.

The only other way would be then to inject a small object with the ability to get the debug backtrace to mock it then, this would split the concerns.
This object could hold the configuration to use debug_backtrace or not as there is other comments about that.

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php (hunk 7 above)
Comment (mitelg) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: adding a config to the shopware.yaml sounds like a good idea :+1:

File: src/Core/Framework/Resources/config/packages/shopware.yaml [hunk 8]
```diff
@@ -561,6 +561,8 @@ shopware:
             delay_options:
                 storage: mysql
             http_cache: ['logged-in', 'cart-filled']
+            log_tags: false
```
Comment (umutdogan4291) [inline src/Core/Framework/Resources/config/packages/shopware.yaml]: Questions:

1- Do we want to combine these two new configs?
2- Do we want to log tags as default? I would keep it false.

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php [hunk 9]
```diff
@@ -106,11 +108,41 @@ private function purge(array $keys): void
...
+                continue; // skip CacheInvalidator methods
+            }
+
+            return [
+                'class' => $frame['class'],
+                'function' => $frame['function'],
+            ];
+        }
+
+        return null;
```
Comment (AydinHassan) [inline src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: Now you implemented a backtrace collector, I would push all this logic in there. Maybe something like `firstFrame(callable $filter): Frame` ? could also just return an array as is now.

File: src/Core/Framework/DependencyInjection/cache.xml [hunk 10]
```diff
@@ -50,6 +50,9 @@
             <argument type="service" id="cache.http"/>
             <argument>%shopware.http_cache.soft_purge%</argument>
             <argument>%shopware.cache.invalidation.delay_enabled%</argument>
+            <argument>%shopware.cache.invalidation.log_tags%</argument>
```
Comment (AydinHassan) [inline src/Core/Framework/DependencyInjection/cache.xml]: we tend to use the `enabled` prefix with toggles, i think it makes sense here also. maybe `tag_invalidation_log_enabled` ?

File: src/Core/Framework/Resources/config/packages/shopware.yaml (hunk 8 above)
Comment (AydinHassan) [reply src/Core/Framework/Resources/config/packages/shopware.yaml]: I think the enabled flag is enough now. Level could still be useful, but when you specifically enable it and parse those records out, you will probably just filter for the message.

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php [hunk 11]
```diff
@@ -106,11 +108,41 @@ private function purge(array $keys): void
             $this->httpCacheStore->setMultiple($list);
         }
 
+        if ($this->logInvalidations) {
+            $this->logger->log(
+                $this->logLevel,
+                \sprintf('Purged %d tags.', \count($keys)),
```
Comment (AydinHassan) [inline src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: Does it make sense to write the message as `Purged tags (%d)...` ? Just so there is something a bit more unique/consistent to search for?

**Review by AydinHassan — CHANGES_REQUESTED** [review]

File: src/Core/Framework/Resources/config/packages/shopware.yaml (hunk 8 above)
Comment (mitelg) [reply src/Core/Framework/Resources/config/packages/shopware.yaml]: yes, on/off toggle should be enough. afaik, we do not allow to configure log levels at other places in that detail :thinking:

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php (hunk 9 above)
Comment (mitelg) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: please no more arrays as return types anymore! use DTO instead
https://peakd.com/php/@crell/php-use-associative-arrays-basically-never
https://presentations.garfieldtech.com/slides-never-use-arrays/phpday2021/

**Review by mitelg — CHANGES_REQUESTED** [review]

File: src/Core/Framework/Util/Dto/BacktraceFrame.php [hunk 12]
```diff
@@ -0,0 +1,15 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\Framework\Util\Dto;
```
Comment (mitelg) [inline src/Core/Framework/Util/Dto/BacktraceFrame.php]: we usually do not use namespaces with naming it what the things are :thinking: or do you found it somewhere like this?

my suggestion would be to create `Shopware\Core\Framework\Util\Backtrace` and put the `TraceCollector` and `Frame` into that

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php [hunk 13]
```diff
@@ -106,6 +107,21 @@ private function purge(array $keys): void
             $this->httpCacheStore->setMultiple($list);
         }
 
+        if ($this->tagInvalidationLogEnabled) {
+            $callerFrame = $this->backtraceCollector->getFirstFrame(
+                fn (array $frame) => empty($frame['class'])
```
Comment (mitelg) [inline src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: please do not use empty. you should explicitly check for what you do not want to have

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php [hunk 14]
```diff
@@ -106,6 +107,21 @@ private function purge(array $keys): void
...
+                    || $frame['class'] === self::class
+            );
+
+            $this->logger->info(
+                \sprintf('Purged tags (%d).', \count($keys)),
+                [
+                    'tags' => $keys,
+                    'caller' => $callerFrame,
```
Comment (mitelg) [inline src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: how does monolog handle the object? probably not a problem, as we only have two strings in it, right?

**Review by mitelg — CHANGES_REQUESTED** [review]

File: src/Core/Framework/Util/Dto/BacktraceFrame.php (hunk 12 above)
Comment (umutdogan4291) [reply src/Core/Framework/Util/Dto/BacktraceFrame.php]: Actually I have seen Util\Exception and I decided to create Dto folder next to it. But I agree that it's easy to find things when they are sorted by domain.

Moved into Util\Backtrace

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php (hunk 14 above)
Comment (umutdogan4291) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: :thinking: we can serialize it in logger->info and keep it safe.

What about this?
```
$this->logger->info(
    \sprintf('Purged tags (%d).', \count($keys)),
    [
        'tags' => $keys,
        'caller' => $callerFrame?->jsonSerialize(),
    ]
);
```

File: src/Core/Framework/Util/Backtrace/Frame.php [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+final readonly class Frame
+{
+    public function __construct(
+        public ?string $class,
+        public ?string $function,
+    ) {
+    }
+
+    /**
+     * @return array<string, ?string>
```
Comment (mitelg) [inline src/Core/Framework/Util/Backtrace/Frame.php]:
```suggestion
     * @return array{class: string|null, function: string|null}
```

if you can, be explicit about the array shape. the more information, the better :+1:

File: src/Core/Framework/Util/Backtrace/Frame.php [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+     */
+    public function jsonSerialize(): array
```
Comment (mitelg) [inline src/Core/Framework/Util/Backtrace/Frame.php]: I would call it `toArray` as nothing is JSON serialized here or at usage

Comment (umutdogan4291) [reply src/Core/Framework/Util/Backtrace/Frame.php]: makes sense. btw thanks for the comments :)

**Review by mitelg — APPROVED** [review]

File: tests/unit/Core/Framework/Adapter/Cache/CacheInvalidatorTest.php [hunk 17]
```diff
@@ -303,9 +407,26 @@ public function testStoreFailureFallsBackToImmediateInvalidation(): void
...
         );
 
         $invalidator->invalidate(['foo']);
     }
+
+    private function createBacktraceCollectorMock(?string $class = null, ?string $function = null): BacktraceCollector
+    {
+        $collector = $this->createMock(BacktraceCollector::class);
+
+        if (!empty($class) && !empty($function)) {
```
Comment (AydinHassan) [inline tests/unit/Core/Framework/Adapter/Cache/CacheInvalidatorTest.php]: @mitelg can't believe you missed this 😄

**Review by AydinHassan — APPROVED** [review]

File: tests/unit/Core/Framework/Adapter/Cache/CacheInvalidatorTest.php (hunk 17 above)
Comment (mitelg) [reply tests/unit/Core/Framework/Adapter/Cache/CacheInvalidatorTest.php]: :scream: 

why you do this to me @umutdogan4291 ???

**Review by mitelg — APPROVED** [review]

File: tests/unit/Core/Framework/Adapter/Cache/CacheInvalidatorTest.php (hunk 17 above)
Comment (umutdogan4291) [reply tests/unit/Core/Framework/Adapter/Cache/CacheInvalidatorTest.php]: looks like a betrayal :sob:

File: src/Core/Framework/Adapter/Cache/CacheInvalidator.php (hunk 6 above)
Comment (umutdogan4291) [reply src/Core/Framework/Adapter/Cache/CacheInvalidator.php]: Hi @nfortier-shopware 

I have made some changes after your comment. Can we resolve this comment now?

**Review by nfortier-shopware — APPROVED** [review]
🚀
